Karl/test.cc

```diff
@@ -322,6 +466,165 @@ TEST(Karl, Deregister) {
+    EXPECT_EQ(200, static_cast<int>(response.code));
+    while (karl.ActiveServicesCount() == 0u) {
+      ;  // Spin lock.
+    }
+    const auto result = karl.InternalExposeStorage()
+                            .ReadOnlyTransaction([&](ImmutableFields<unittest_karl_t::storage_t> fields) {
+                              ASSERT_TRUE(Exists(fields.claires[claire.codename]));
+                              EXPECT_EQ(current::karl::ClaireRegisteredState::Active,
+                                        Value(fields.claires[claire.codename]).registered_state);
+                            })
+                            .Go();
+    EXPECT_TRUE(WasCommitted(result));
+  }
+
+  current::time::SetNow(std::chrono::microseconds(100 * 1000 * 1000),
+                        std::chrono::microseconds(101 * 1000 * 1000));
+  while (karl.ActiveServicesCount() > 0u) {
+    ;  // Spin lock.
+  }
+  {
+    const auto result = karl.InternalExposeStorage()
+                            .ReadOnlyTransaction([&](ImmutableFields<unittest_karl_t::storage_t> fields) {
+                              ASSERT_TRUE(Exists(fields.claires[claire.codename]));
+                              EXPECT_EQ(current::karl::ClaireRegisteredState::DisconnectedByTimeout,
+                                        Value(fields.claires[claire.codename]).registered_state);
+                            })
+                            .Go();
+    EXPECT_TRUE(WasCommitted(result));
+  }
+}
+
+TEST(Karl, DisconnectedByTimoutWithNginx) {
+  current::time::ResetToZero();
+
+  // Run the test only if `karl_nginx_config_file` flag is set.
+  if (FLAGS_karl_nginx_config_file.empty()) {
+    return;
+  }
+
+  const auto stream_file_remover = current::FileSystem::ScopedRmFile(FLAGS_karl_test_stream_persistence_file);
+  const auto storage_file_remover = current::FileSystem::ScopedRmFile(FLAGS_karl_test_storage_persistence_file);
+
+  const current::karl::KarlNginxParameters nginx_parameters(FLAGS_karl_nginx_port,
+                                                            FLAGS_karl_nginx_config_file);
+  const std::string karl_nginx_base_url = "http://localhost:" + current::ToString(FLAGS_karl_nginx_port);
+  unittest_karl_t karl(FLAGS_karl_test_port,
+                       FLAGS_karl_test_stream_persistence_file,
+                       FLAGS_karl_test_storage_persistence_file,
+                       "/",
+                       karl_nginx_base_url,
+                       nginx_parameters);
+  const current::karl::Locator karl_locator(Printf("http://localhost:%d", FLAGS_karl_test_port));
+
+  current::karl::ClaireStatus claire;
+  claire.service = "unittest";
+  claire.codename = "ABCDEF";
+  claire.local_port = PickPortForUnitTest();
+  // Register fake service.
+  HTTP(claire.local_port).Register("/.current", [](Request r) { r("GOTIT\n"); });
+
+  std::string proxied_url;
+  {
+    const std::string keepalive_url = Printf("%s?codename=%s&port=%d",
```

Comment:
Why not declare and initialize `current::karl::ClaireStatus claire;` in a very narrow scope running the POST on the next line, and use a bare `"ABCDEF`" string everywhere else? No need to DRY the test.